spec/abstract/AARCH64/VCPUAcc_A.thy

```diff
@@ -263,6 +263,37 @@ definition vcpu_switch :: "obj_ref option \<Rightarrow> (unit,'z::state_ext) s_m
               od))
      od"
 
+text \<open>
+  Flush the current VCPU if there is one, saving any existing state before disabling VCPU mode and
+  setting it so that there is no current VCPU.\<close>
+definition vcpu_flush :: "(unit,'z::state_ext) s_monad" where
+  "vcpu_flush \<equiv> do
+     cur_v \<leftarrow> gets (arm_current_vcpu \<circ> arch_state);
+     when (cur_v \<noteq> None) $ do
+       vcpu_save cur_v;
+       vcpu_invalidate_active
+     od
+   od"
+
+text \<open>Flush the given TCB's VCPU if it is the current VCPU.\<close>
+definition vcpu_flush_if_current :: "obj_ref \<Rightarrow> (unit,'z::state_ext) s_monad" where
+  "vcpu_flush_if_current t \<equiv> do
+     cur_vcpu \<leftarrow> gets (arm_current_vcpu \<circ> arch_state);
+     vcpu \<leftarrow> arch_thread_get tcb_vcpu t;
+     when (vcpu = map_option fst cur_vcpu) vcpu_flush
+   od"
+
+(*FIXME: find the right place to move this to, ArchTcb_A doesn't work as it is imported by VCPUAcc_A
+    and leads to an unresolvable circular dependency*)
+definition arch_prepare_set_domain :: "obj_ref \<Rightarrow> domain \<Rightarrow> (unit,'z::state_ext) s_monad" where
```

Comment:
I don't think it is a problem that the function is here in HYP architectures and in `ArchTcb_A` in non-HYP architectures. I'd put a comment in the non-HYP architectures that it is in `ArchTcb_A`, because they don't have the `VCPUAcc_A` file to explain the discrepancy, and maybe leave comment here that it can't go into `ArchTCB_A`, because that would lead to a circular dependency. Then we know what's going on and can find things from either direction.